File: backend/src/Solitaire.Engine/Cards.cs
namespace Solitaire.Engine;

/// <summary>
/// The four French-deck suits. The integer values are part of the engine's
/// wire/deterministic contract and MUST match the TypeScript port:
/// Clubs = 0, Diamonds = 1, Hearts = 2, Spades = 3.
/// </summary>
public enum Suit
{
    Clubs = 0,
    Diamonds = 1,
    Hearts = 2,
    Spades = 3,
}

/// <summary>Card color, derived from <see cref="Suit"/>.</summary>
public enum CardColor
{
    Black = 0,
    Red = 1,
}

/// <summary>
/// A single playing card. <see cref="Rank"/> is 1..13 (Ace = 1, Jack = 11,
/// Queen = 12, King = 13).
/// </summary>
public readonly record struct Card(Suit Suit, int Rank)
{
    /// <summary>Ace rank constant.</summary>
    public const int Ace = 1;

    /// <summary>King rank constant.</summary>
    public const int King = 13;

    /// <summary>Red suits are Diamonds and Hearts; Clubs and Spades are black.</summary>
    public CardColor Color => Suit is Suit.Diamonds or Suit.Hearts ? CardColor.Red : CardColor.Black;

    /// <summary>
    /// Canonical 0..51 index used to build the ordered deck before shuffling.
    /// Suit-major: <c>index = (int)Suit * 13 + (Rank - 1)</c>. This ordering is
    /// part of the deterministic contract and must match the TypeScript port.
    /// </summary>
    public int OrdinalIndex => (int)Suit * 13 + (Rank - 1);

    /// <summary>Builds a card from its canonical 0..51 <see cref="OrdinalIndex"/>.</summary>
    public static Card FromOrdinalIndex(int index)
    {
        if (index is < 0 or > 51)
        {
            throw new ArgumentOutOfRangeException(nameof(index), index, "Card index must be 0..51.");
        }

        return new Card((Suit)(index / 13), (index % 13) + 1);
    }

    /// <summary>Short debug form, e.g. "AS", "TD", "KH", "2C".</summary>
    public override string ToString()
    {
        char rank = Rank switch
        {
            1 => 'A',
            10 => 'T',
            11 => 'J',
            12 => 'Q',
            13 => 'K',
            _ => (char)('0' + Rank),
        };
        char suit = Suit switch
        {
            Suit.Clubs => 'C',
            Suit.Diamonds => 'D',
            Suit.Hearts => 'H',
            Suit.Spades => 'S',
            _ => '?',
        };
        return $"{rank}{suit}";
    }
}

File: frontend/src/engine/cards.ts
// Card model — a direct port of Cards.cs. The integer suit values and the
// suit-major ordinal are part of the deterministic contract shared with the C#
// engine and must not change.

export enum Suit {
  Clubs = 0,
  Diamonds = 1,
  Hearts = 2,
  Spades = 3,
}

export const ACE = 1;
export const KING = 13;

export interface Card {
  readonly suit: Suit;
  readonly rank: number; // 1..13 (Ace = 1, King = 13)
}

/** Diamonds and Hearts are red; Clubs and Spades are black. */
export function isRed(suit: Suit): boolean {
  return suit === Suit.Diamonds || suit === Suit.Hearts;
}

/** True if two suits share a color (used for alternating-color rules). */
export function sameColor(a: Suit, b: Suit): boolean {
  return isRed(a) === isRed(b);
}

/** Canonical 0..51 index: suit-major, `suit * 13 + (rank - 1)`. */
export function ordinalIndex(card: Card): number {
  return card.suit * 13 + (card.rank - 1);
}

/** Builds a card from its canonical 0..51 ordinal index. */
export function cardFromOrdinal(index: number): Card {
  return { suit: Math.floor(index / 13), rank: (index % 13) + 1 };
}
